### PR TITLE
executor: add deprecation warning for hash_join_concurrency

### LINKED ISSUE #15428: UCP: Control the concurrency of all types of executors in one session variable
## Description

Currently, TiDB has several variables to control each type of executor,

See https://github.com/pingcap/tidb/blob/45d83d1adb3eedc83bf5f59f21e6666dfcf8696f/sessionctx/variable/session.go#L1282-L1309

And each of them has its own default value.
I propose to add one session variable named like`ExecutorConcurrency` to control the concurrency of all types of executors and make the default value as the number of the CPU cores.

For compatibility of the old variables, take `HashJoinConcurrency` as an example,
1) The default value of `ExecutorConcurrency` is  the number of the CPU cores.
2) Remove the default value of `HashJoinConcurrency`. 
3) If `HashJoinConcurrency` is unset, the concurrency of Hash Join is applied by `ExecutorConcurrency`. If `select @tidb_hash_join_concurrency;` is queried, the value of `ExecutorConcurrency` is outputted.
4) If `HashJoinConcurrency` is set specifically,  the concurrency of Hash Join is applied by `HashJoinConcurrency`. If `select @tidb_hash_join_concurrency;` is queried, the value of `HashJoinConcurrency` is outputted.

## Score

* 300

## Mentor(s)

* @SunRunAway 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang

## Learning Materials

* Chinese: [TiDB 精选技术讲解文章](https://github.com/pingcap/presentations/blob/master/hackathon-2019/reference-document-of-hackathon-2019.md)
* English: [Awesome-Database-Learning](https://github.com/pingcap/awesome-database-learning)

### PR DESCRIPTION
**Pull request by niedhui**
### What problem does this PR solve?

Issue Number: close #15428

Problem Summary: 
This should be the last PR for #15428, add deprecation warning when setting `tidb_hash_join_concurrency` 

### What is changed and how it works?

What's Changed:

Introduce a function called  `CheckDeprecationSetSystemVar` to handle deprecation warnings when setting system variables, it is used in `SetGlobalSysVar` and `SetSessionSystemVar` (deprecation warning was only added for `SetSessionSystemVar` before) .

### Check List 

- Integration test
- Manual test (add detailed scripts or steps below)

### Release note 

No release note

### CODE REVIEW

**Comment by niedhui** [comment]
@sre-bot /run-all-tests

**Comment by niedhui** [comment]
Fail tests should be passed after https://github.com/pingcap/mysql-tester/pull/5 is merged

File: executor/set.go [hunk 1]
```diff
@@ -203,13 +211,13 @@ func (e *SetExecutor) setSysVariable(name string, v *expression.VarAssignment) e
...
-		if name != variable.AutoCommit {
-			logutil.BgLogger().Info("set session var", zap.Uint64("conn", sessionVars.ConnectionID), zap.String("name", name), zap.String("val", valStr))
-		} else {
-			// Some applications will set `autocommit` variable before query.
-			// This will print too many unnecessary log info.
-			logutil.BgLogger().Debug("set session var", zap.Uint64("conn", sessionVars.ConnectionID), zap.String("name", name), zap.String("val", valStr))
-		}
+	}
+	if name != variable.AutoCommit {
+		logutil.BgLogger().Info(fmt.Sprintf("set %s var", scopeStr), zap.Uint64("conn", sessionVars.ConnectionID), zap.String("name", name), zap.String("val", valStr))
```
Comment (SunRunAway) [inline executor/set.go:216]: Does this change close #17091?

Comment (niedhui) [reply executor/set.go:216]: Yeah, I think so.
```
mysql> set @@tidb_hash_join_concurrency=2;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> set @@global.tidb_hash_join_concurrency=2;
Query OK, 0 rows affected, 1 warning (0.00 sec)
```


```
[2020/07/09 11:11:50.981 +08:00] [INFO] [set.go:216] ["set session var"] [conn=1] [name=tidb_hash_join_concurrency] [val=2]
[2020/07/09 11:12:00.222 +08:00] [INFO] [set.go:216] ["set global var"] [conn=1] [name=tidb_hash_join_concurrency] [val=2]

```

**Comment by niedhui** [comment]
@sre-bot /run-all-tests

**Comment by niedhui** [comment]
@sre-bot /run-common-test

**Comment by niedhui** [comment]
Ah, tests fail again, @SunRunAway PTAL
The formatted error message is: 
```
run test [role] err: sql:SELECT * FROM mysql.role_edges order by FROM_HOST;: failed to run query
"SELECT * FROM mysql.role_edges order by FROM_HOST;"
 around line 154,
we need(268):
SELECT * FROM mysql.role_edges order by FROM_HOST;
FROM_HOST       FROM_USER       TO_HOST TO_USER WITH_ADMIN_OPTION
%       consultants     %       joan    N
%       consultants     %       sally   N
%       engineering     %       sally   N
%       qa      %       consultants     N
%       wp_administrators       localhost       joe     N
US      engineering     INDIA   engineering     N

but got(268):
SELECT * FROM mysql.role_edges order by FROM_HOST;
FROM_HOST       FROM_USER       TO_HOST TO_USER WITH_ADMIN_OPTION
%       engineering     %       sally   N
%       consultants     %       joan    N
%       consultants     %       sally   N
%       qa      %       consultants     N
%       wp_administrators       localhost       joe     N
US      engineering     INDIA   engineering     N
```

File: executor/set.go (hunk 1 above)
Comment (SunRunAway) [reply executor/set.go:216]: @niedhui Would you like to help put this change in an individual pull request so it can be cherry-picked to older release?

Comment (niedhui) [reply executor/set.go]: done. https://github.com/pingcap/tidb/pull/18457

**Comment by niedhui** [comment]
@sre-bot /run-all-tests

**Comment by niedhui** [comment]
@sre-bot /run-common-test

**Comment by SunRunAway** [comment]
@ti-srebot /run-all-tests

**Comment by niedhui** [comment]
@ti-srebot /run-all-tests

**Pull request merged**